### PR TITLE
NAS for TensorFlow

### PR DESCRIPTION
**Pull request by liuzhe-lz**
I will re-generate the commit history to make this patch clear. Expected commit order is shown as follow.
The items in _itatic_ font are not yet ready.
Some features may be moved to future PRs.

+ [x] mutable interface
+ [x] mutator interface
+ [x] ENAS
+ [ ] ENAS-specific static graph
+ [ ] DARTS
+ [ ] DARTS-specific static graph
+ [ ] docstring
+ [ ] _Proxyless NAS_ (under evaluation)
+ [ ] _General static graph interface_
+ [ ] _Port static graph interface to model compression_
+ [ ] _Other NAS algorithms_

### CODE REVIEW

File: src/sdk/pynni/nni/nas/tensorflow/mutables.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+import logging
+
+from tensorflow.keras import Model
+
+
+_logger = logging.getLogger(__name__)
+
+_counter = 0
+
+def _global_mutable_counting():
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/nas/tensorflow/mutables.py]: this counting makes it hard to do key matching of two independent running. For example, one is training and the other is applying.

**Comment by QuanluZhang** [comment]
better to add docstring

**Comment by ultmaster** [comment]
These codes look almost identical to those of PyTorch. Can we have some diffchecker that can diff tensorflow directory with pytorch directory? To me, another draft PR with overwriting would be great.

File: src/sdk/pynni/nni/nas/tensorflow/base_mutator.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        return root
+
+    @property
+    def mutables(self):
+        return self._structured_mutables
+
+    def undedup_mutables(self):
+        return self._structured_mutables.traverse(deduplicate=False)
+
+    def forward(self, *inputs):
```
Comment (ultmaster) [inline src/sdk/pynni/nni/nas/tensorflow/base_mutator.py]: I think forward is not needed for tensorflow.

File: src/sdk/pynni/nni/nas/tensorflow/mutables.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+class LayerChoice(Mutable):
+    def __init__(self, op_candidates, reduction='sum', return_mask=False, key=None):
+        super().__init__(key=key)
+        self.length = len(op_candidates)
+        self.choices = op_candidates
+        self.reduction = reduction
+        self.return_mask = return_mask
+
+    def forward(self, *inputs):
```
Comment (ultmaster) [inline src/sdk/pynni/nni/nas/tensorflow/mutables.py]: `call`?

File: src/sdk/pynni/nni/nas/tensorflow/mutables.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        assert n_chosen is None or 0 <= n_chosen <= n_candidates, \
+                'Expected selected number must be None or no more than number of candidates.'
+
+        self.n_candidates = n_candidates
+        self.choose_from = choose_from.copy()
+        self.n_chosen = n_chosen
+        self.reduction = reduction
+        self.return_mask = return_mask
+
+    def forward(self, optional_inputs):
```
Comment (ultmaster) [inline src/sdk/pynni/nni/nas/tensorflow/mutables.py]: same here

File: src/sdk/pynni/nni/nas/tensorflow/enas/trainer.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+skip_weight = 0.8
+baseline_decay = 0.999
+child_steps = 500
+mutator_lr = 0.00035
+mutator_steps_aggregate = 20
+mutator_steps = 50
+test_arc_per_epoch = 1
+
+
+class EnasTrainer:
```
Comment (ultmaster) [inline src/sdk/pynni/nni/nas/tensorflow/enas/trainer.py]: Is there an example ready for testing this trainer from end to end?

**Comment by ultmaster** [comment]
If no further improvement is planned for this PR, I'm strongly against merging it into master. Merging into a dev branch might be more acceptable.

**Pull request merged**